src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs

```diff
@@ -86,7 +90,39 @@ protected Task<ImmutableArray<CodeActionOperation>> GetPreviewOperationsAsync(Ca
                 async () => await CodeAction.GetPreviewOperationsAsync(cancellationToken).ConfigureAwait(false), cancellationToken);
         }
 
-        public virtual void Invoke(CancellationToken cancellationToken)
+        public void Invoke(CancellationToken cancellationToken)
+        {
+            this.AssertIsForeground();
+            
+            // Create a task to do the actual async invocation of this action.
+            // For testing purposes mark that we still have an outstanding async 
+            // operation so that we don't try to validate things too soon.
+            var asyncToken = OperationListener.BeginAsyncOperation(GetType().Name + "." + nameof(Invoke));
+            var task = YieldThenInvokeAsync(cancellationToken);
+            task.CompletesAsyncOperation(asyncToken);
+        }
+
+        private async Task YieldThenInvokeAsync(CancellationToken cancellationToken)
+        {
+            this.AssertIsForeground();
+
+            // Yield the UI thread so that the light bulb can be dismissed.  This is necessary
+            // as some code actions may be long running, and we don't want the light bulb to
+            // stay on screen.
+            await Task.Yield();
```

Comment:
We immediately pop up the threaded wait dialog.  So reentrancy seems very unlikely.  Specifically:

The command is invoked.  That's going to be on the UI thread.  Then we yield and make the TWD.  So hte only way to get reentrancy would be if somehow there was a queued message after the invoke.   

1. that seems incredibly unlikely/difficult to get into.
2. i can't even think of what sort of message we might get that would be problematic.

Note: the experience without this change is exceedingly poor.